pilot/pkg/serviceregistry/serviceentry/servicediscovery.go

```diff
@@ -343,6 +343,14 @@ func (s *ServiceEntryStore) serviceEntryHandler(old, curr config.Config, event m
 
 // WorkloadInstanceHandler defines the handler for service instances generated by other registries
 func (s *ServiceEntryStore) WorkloadInstanceHandler(wi *model.WorkloadInstance, event model.Event) {
+
+	// we only update the service registry in the below conditions:
```

Comment:
Sorry I didn't realize this before, but it seems this PR is trying to solve two issues together:
1. We send EDS endpoints for unhealthy pods (bug)
2. We send EDS updates when there is no real change (ie readiness didn't change) - performance optimization

What I am a bit confused about is why we need InTransitState. It seems its only for (2), but don't we already solve this by `redundantEventForPod` a few lines down?

I am a bit confused why this PR needs any of the "transit" stuff instead of just ensuring that unready instances are not sent, which would be a much simpler change?

Reply:
@howardjohn The (2) is a side effect of (1). 

The reason we need `InTransitState` is because there are two types of unreadiness. 

1. Pods that are being created and becoming ready => we need to ignore the unreadiness events and only push once it has become ready
2. Pods that are ready and becoming unready => we need to push(delete) once it has become unready

Without `InTransitState` flag, there is no way to differentiate these two types of unreadiness.

Reply:
I am not sure why (1) is not handled by `redundantEventForPod`. It would come in as a delete event, and we hit `!exists` since the its new and skip it. The registry has enough info to know if it needs to push without additional InTransitState metadata (as "proof", k8s endpoints controller is doing the ~same thing and doesn't look at this field).

Reply:
I don't think in the case of (1) there would be any `delete` event. It should start with an `add` event and followed by `update` event. I just did a test on 1.7.3, I can clearly see an unready pod being added to the service entry.

Reply:
@howardjohn The suggestion drives me do some testing locally. Using `EventDelete` seems to be working and it's passing the tests. Do you have any concern for other handlers that might rely on the event type? Because in the first PR you mentioned that we have two types of push relying on this part of code: 1) endpoints to sidecar as upstream 2) config to the pod's sidecar

Does this will break 2)?

```		case model.EventAdd:
			switch pod.Status.Phase {
			case v1.PodPending, v1.PodRunning:
				if key != pc.podsByIP[ip] {
					// add to cache if the pod is running or pending
					pc.update(ip, key)
				}
				if !IsPodReady(pod) {
					ev = model.EventDelete // <= not sure if line will break 2)
				}
			}
```